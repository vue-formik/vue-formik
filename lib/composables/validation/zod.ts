import { ZodType } from "zod";

import { setNestedValue } from "@/helpers";
import { constructPath } from "@/helpers/formikObject/objectPath";

const validateZod = <T extends object>(values: T, schema: ZodType<Partial<T>>) => {
  const errors = {};
  const result = schema.safeParse(values);
  if (!result.success) {
    result.error.errors.forEach(({ path, message }) => {
      setNestedValue(errors as Record<string, unknown>, constructPath(path), message);
    });
  }
  return errors;
};

export default validateZod;
