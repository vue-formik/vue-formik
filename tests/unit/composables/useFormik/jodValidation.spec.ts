import { describe, test, expect } from "vitest";
import { z } from "zod";
import { useFormik } from "@/index";
import { nextTick } from "vue";

describe("useFormik/jodValidation", async () => {

  test("simple form validation", async () => {
    const initialValues = {
      name: "",
      email: "",
    }
    const validationSchema = z.object({
      name: z.string().min(3),
      email: z.string().email(),
    })

    const formik = useFormik({
      initialValues,
      validationSchema,
      validateOnMount: true,
      mode: "JOD"
    })

    expect(formik.errors).toMatchSnapshot()

    formik.setFieldValue("name", "John")
    formik.setFieldValue("email", "john@ex.com")
    await nextTick()
    expect(formik.errors).toMatchSnapshot()
    expect(formik.isValid.value).toBe(true)
  })

  test("form with array fields", async () => {
    const initialValues = {
      name: "",
      email: "",
      friends: [
        { name: "" },
        { name: "" },
      ]
    }
    const validationSchema = z.object({
      name: z.string().min(3),
      email: z.string().email(),
      friends: z.array(z.object({
        name: z.string().min(3),
      }))
    })

    const formik = useFormik({
      initialValues,
      validationSchema,
      validateOnMount: true,
      mode: "JOD"
    })

    expect(formik.errors).toMatchSnapshot()

    formik.setFieldValue("name", "John")
    formik.setFieldValue("email", "john@ex.com")
    formik.setFieldValue("friends[0].name", "Jane")
    formik.setFieldValue("friends[1].name", "Doe")
    await nextTick()

    expect(formik.errors).toMatchSnapshot()
    expect(formik.isValid.value).toBe(true)

    formik.setFieldValue("friends[0].name", "J")
    await nextTick()

    expect(formik.errors).toMatchSnapshot()
    expect(formik.isValid.value).toBe(false)
  })
})
