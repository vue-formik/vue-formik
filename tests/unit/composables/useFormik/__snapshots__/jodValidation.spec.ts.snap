// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFormik/jodValidation > form with array fields 1`] = `
{
  "email": "Invalid email",
  "friends": {
    "0": {
      "name": "String must contain at least 3 character(s)",
    },
    "1": {
      "name": "String must contain at least 3 character(s)",
    },
  },
  "name": "String must contain at least 3 character(s)",
}
`;

exports[`useFormik/jodValidation > form with array fields 2`] = `{}`;

exports[`useFormik/jodValidation > form with array fields 3`] = `
{
  "friends": {
    "0": {
      "name": "String must contain at least 3 character(s)",
    },
  },
}
`;

exports[`useFormik/jodValidation > simple form validation 1`] = `
{
  "email": "Invalid email",
  "name": "String must contain at least 3 character(s)",
}
`;

exports[`useFormik/jodValidation > simple form validation 2`] = `{}`;
